lib/xander/transaction.ex

```diff
@@ -0,0 +1,312 @@
+        :establish,
+        %__MODULE__{client: client, socket: socket, network: network} = data
+      ) do
+    Logger.info("Establishing handshake...")
+
+    :ok =
+      client.send(
+        socket,
+        Handshake.Proposal.version_message(@active_n2c_versions, network)
+      )
+
+    case client.recv(socket, 0, _timeout = 5_000) do
+      {:ok, full_response} ->
+        case Handshake.Response.validate(full_response) do
+          {:ok, _handshake_response} ->
+            # Transitions to idle state
+            Logger.info("Handshake successful")
+            {:next_state, :idle, data}
+
+          {:refused, response} ->
+            Logger.error("Handshake refused by node. Response: #{inspect(response)}")
+
+            {:next_state, :disconnected, data}
+        end
+
+      {:error, reason} ->
+        Logger.error("Error establishing connection #{inspect(reason)}")
+        {:next_state, :disconnected, data}
+    end
+  end
+
+  @spec busy(
+          :info | :internal | {:call, any()},
+          {:request, atom(), atom()} | {:tcp_closed, any()},
+          t()
+        ) ::
+          {:next_state, :disconnected, t()}
+          | {:next_state, :disconnected | :idle, t(), [{:reply, any(), {:error | :ok, any()}}]}
+          | {:keep_state, t(), [{:next_event, :internal, :send_tx}]}
+  def busy(_event_type, _event_content, _data)
+
+  # Handle call from dependent process to send a transaction. The transaction and caller are
+  # added to the queue, and will be processed after the current transaction response is received.
+  def busy(
+        {:call, from},
+        {:request, :send_tx, tx_hash},
+        %__MODULE__{client: _client, socket: _socket} = data
+      ) do
+    # Add the caller and transaction in our queue
+    data = update_in(data.queue, &:queue.in({from, tx_hash}, &1))
+    {:keep_state, data, [{:next_event, :internal, :send_tx}]}
+  end
+
+  # Handle internal event to send a transaction (from idle state)
+  def busy(
+        :internal,
+        :send_tx,
+        %__MODULE__{client: client, socket: socket} = data
+      ) do
+    # Get the current transaction from queue without removing it
+    {:value, {_from, tx_hash}} = :queue.peek(data.queue)
+
+    # Set socket to active mode before sending to be ready to receive response
+    :inet.setopts(socket, active: true)
+
+    # Send transaction to node and remain in busy state
+    :ok = client.send(socket, Messages.transaction(tx_hash))
+
+    {:keep_state, data}
+  end
+
+  def busy(:info, {:tcp_closed, socket}, %__MODULE__{socket: socket} = data) do
+    Logger.error("Connection closed while in busy")
+    {:next_state, :disconnected, data}
+  end
+
+  # This function is invoked due to the socket being set to active mode.
+  # Handles the response while in busy state.
+  def busy(
+        :info,
+        {_tcp_or_ssl, socket, bytes},
+        %__MODULE__{client: _client, socket: socket} = data
+      ) do
+    handle_response(bytes, socket, data)
+  end
+
```

Comment:
We need just one more clause for the `busy` function to match on `{:tcp_error, socket, reason}`. See https://www.erlang.org/doc/apps/kernel/gen_tcp.html#connect/4-socket-data

Reply:
nice, added.